### PR TITLE
Add a reference counter class for intrusive ptrs.

### PR DESCRIPTION
**Pull request by bgogul**
Adds a counter class for reference counting in an intrusive ptr implementation.

### CODE REVIEW

File: src/utils/ref_counter.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#define SRC_UTILS_REF_COUNTER_H_
+
+#include <atomic>
+#include <cstddef>
+
+#include "src/common/logging/logging.h"
+
+namespace raksha {
+
+template <bool thread_safe>
```
Comment (markww) [inline src/utils/ref_counter.h:26]: It looks like the thread safety isn't being used right now; is it necessary?

File: src/utils/ref_counter.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+class RefCounter {
+ public:
+  RefCounter() : count_(0) {}
+  RefCounter(size_t init) : count_(init) {}
+
+  operator size_t() const { return count_; }
+
+  // Adds the given increment and returns the old value.
+  size_t FetchAdd(size_t increment,
+                  std::memory_order order = std::memory_order_seq_cst) {
```
Comment (markww) [inline src/utils/ref_counter.h:36]: Is sequential consistency necessary, or can we get away with acqrel? Also, do we need this when we haven't implemented the atomicity portion?

File: src/utils/ref_counter.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    size_t old = count_;
+    count_ += increment;
+    return old;
+  }
+
+  // Subtracts the given increment and returns the old value.
+  size_t FetchSub(size_t decrement,
+                  std::memory_order order = std::memory_order_seq_cst) {
```
Comment (markww) [inline src/utils/ref_counter.h:44]: See other question about memory order param.

File: src/utils/ref_counter.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    CHECK(count_ >= decrement) << "FetchSub can decrement below zero.";
+    size_t old = count_;
+    count_ -= decrement;
+    return old;
+  }
+
+ private:
+  // TODO: Create thread-safe version.
+  mutable size_t count_;
```
Comment (markww) [inline src/utils/ref_counter.h:53]: Why is this `mutable`? Are we expecting to manipulate it through a `const` pointer?

File: src/utils/ref_counter.h (hunk 1 above)
Comment (bgogul) [reply src/utils/ref_counter.h:26]: Just wanted to future proof the interface. I guess it is not hard to add this later. Will remove and simplify.

File: src/utils/ref_counter.h (hunk 2 above)
Comment (bgogul) [reply src/utils/ref_counter.h:36]: Not used at all. Just a place holder when we add atomicity.

File: src/utils/ref_counter.h (hunk 3 above)
Comment (bgogul) [reply src/utils/ref_counter.h:44]: Just a place holder. Will remove and simplify.

File: src/utils/ref_counter.h (hunk 4 above)
Comment (bgogul) [reply src/utils/ref_counter.h:53]: Actually, this is not needed here.

**Comment by bgogul** [comment]
@markww, simplified this. PTAL.

**Review by markww — APPROVED** [review]
LGTM

**Pull request merged**